### PR TITLE
#159965315 add feature for authors to tag their articles

### PR DESCRIPTION
**Pull request by meshack-mbuvi**
#### What does this pull request do?
- Enables authors to tag their articles when creating and updating them.

#### Description of tasks to be completed
- Ensure authors can tag their articles when creating or updating them.
- Ensure authors can remove article tags.

#### How should this be manually tested
- Clone [this](https://github.com/andela/ah-magnificent6.git) repository
- Create and activate a [virtual environment](https://docs.python.org/3/library/venv.html) inside the project base directory.
- Run the command `git checkout ft-article-tags-159965315`
- Install the project dependencies by running `pip install -r requirements.txt`
- Execute `python manage.py makemigrations` followed by `python manage.py migrate`
- Run the server by issuing the command `python manage.py runserver`
- Using ([Postman](https://www.getpostman.com/) or your favorite REST API client, 
- Create an account by accessing `api/users/signup/` by sending the payload in the format 
`
{
  "firstname": "admin",
  "username": "admin",
  "email": "admin@gmail.com",
  "password": "password",
} 
`
and 
login on `/api/users/login/` using the same format for the payload above
- test the above endpoints by passing data in the format 
`{
    'title':'title of the article',
    'body':'body of the article',
    'description':'Description of the article',
    'tags': 'A comma seperated sentence'
}`

#### What are the relevant pivotal tracker stories?
[#159965306 User should be able to create articles](https://www.pivotaltracker.com/story/show/159965315)
#### Screenshots 
image

### CODE REVIEW

File: authors/apps/articles/views.py [hunk 1]
```diff
@@ -18,9 +18,33 @@
...
+    # also, add them to the articles and save the article instance
+
+    for tag in tags.split(','):
+        try:
+            data = {'tag': tag.strip()}
+            serializer = TagsSerializer(data=data)
+            serializer.is_valid(raise_exception=True)
+            serializer.save()
+        except:
+            pass
```
Comment (micahoriaso) [inline authors/apps/articles/views.py]: @meshack-mbuvi it is highly discouraged to allow silent exceptions. Please catch this exception and handle it.

File: authors/apps/articles/views.py [hunk 2]
```diff
@@ -58,6 +82,12 @@ def post(self, request):
         serializer = self.serializer_class(data=article)
         serializer.is_valid(raise_exception=True)
         serializer.save()
```
Comment (micahoriaso) [inline authors/apps/articles/views.py]: @meshack-mbuvi here you can save the serializer/model instance on save if you need to use it after a save is performed. For instance:
`saved_article = serializer.save()`
You can then use the `saved_article` object in the `create_tag` method in line 90 of articles/views.py.
i.e `create_tag(tags, saved_article)`
This approach means that we don't have to query the database for the saved article as you have done in line 89

Comment (micahoriaso) [reply authors/apps/articles/views.py]: This approach can also be used in the definition of the `create_tag` method where you have done `new_tag = ArticleTags.objects.get(tag=tag.strip())`

**Review by micahoriaso — CHANGES_REQUESTED** [review]

File: authors/apps/articles/views.py (hunk 2 above)
Comment (meshack-mbuvi) [reply authors/apps/articles/views.py]: Thanks very much. I have learned something from this.

Comment (meshack-mbuvi) [reply authors/apps/articles/views.py]: Check the new changes and if you see anything I can improve, please let me know; it really helps a lot.

**Review by micahoriaso — APPROVED** [review]

File: authors/apps/articles/views.py (hunk 2 above)
Comment (micahoriaso) [reply authors/apps/articles/views.py]: @meshack-mbuvi great work resolving the issues. This PR now looks good to me.

File: authors/apps/articles/views.py (hunk 1 above)
Comment (meshack-mbuvi) [reply authors/apps/articles/views.py]: This one has been resolved as we have discussed.

File: authors/apps/articles/views.py (hunk 2 above)
Comment (meshack-mbuvi) [reply authors/apps/articles/views.py]: This is very helpful.

File: authors/apps/articles/tests/test_article_likes_and_dislikes.py [hunk 3]
```diff
@@ -45,6 +45,22 @@ def test_cannot_like_an_article_more_than_once(self):
                                     )
         self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
 
+    def test_cannot_dislike_an_article_more_than_once(self):
```
Comment (parseen254) [inline authors/apps/articles/tests/test_article_likes_and_dislikes.py:48]: @meshack-mbuvi is this in the scope of your task?

**Review by parseen254 — CHANGES_REQUESTED** [review]
Also, make sure to squash your migrations. They are too many. Recommended resource: https://stackoverflow.com/questions/40028586/how-to-squash-recent-django-migrations

File: authors/apps/articles/tests/test_article_likes_and_dislikes.py (hunk 3 above)
Comment (parseen254) [reply authors/apps/articles/tests/test_article_likes_and_dislikes.py:48]: Also, make sure test coverage is not dropping.

**Comment by Bileonaire** [comment]
#LGTM.....Looks good to me, good work on implementing the changes requested by @micahoriaso.

File: authors/apps/articles/tests/test_article_likes_and_dislikes.py (hunk 3 above)
Comment (meshack-mbuvi) [reply authors/apps/articles/tests/test_article_likes_and_dislikes.py:48]: I wanted to improve test coverage. I always want to achieve a 100% test coverage on everything I work on.

Comment (meshack-mbuvi) [reply authors/apps/articles/tests/test_article_likes_and_dislikes.py:48]: I have made the requested changes.

**Pull request merged**